### PR TITLE
Include image ID and version in environment variables

### PR DESCRIPTION
**Pull request by mcassaniti**
This change will allow the image ID and image version arguments to be referenced in scripts (prepare, build, postinst, finalize). This is useful for substituting or adding into relevant files.

For example:

```shell
$ cat mkosi.postinst
#!/bin/sh

...

echo "IMAGE_ID=$IMAGE_ID"           >> /etc/os-release
echo "IMAGE_VERSION=$IMAGE_VERSION" >> /etc/os-release

# These files were added from mkosi.extra
for f in $(find /usr/lib/sysupdate.d/ -name '*.conf') ; do
    sed --in-place s/'IMAGE_ID'/"$IMAGE_ID"/ $f
done
```

### CODE REVIEW

File: mkosi/backend.py
Comment (behrmann) [inline mkosi/backend.py:701]: Aren't the environment variable names missing here?
```suggestion
    if args.image_id:
        nspawn += [f"--setenv=IMAGE_ID={args.image_id}"]

    if args.image_version:
        nspawn += [f"--setenv=IMAGE_VERSION={args.image_version}"]
```

Comment (mcassaniti) [reply mkosi/backend.py]: 🤦‍♂️ Whoops. Fixed now.

**Review by behrmann — APPROVED** [review]

**Comment by mcassaniti** [comment]
@behrmann Who does the merging once this is approved?

**Comment by behrmann** [comment]
> @behrmann Who does the merging once this is approved?

Anybody in the maintainer's group, as a rule of thumb those are people with the *member* pill next to their name. My usual personal policy is to merge stuff that two people have looked at, i.e. approved or said their LGTM to.

**Comment by behrmann** [comment]
Not sure the errors qemu spews are connected to this, since at least for Debian I think that nothing should pick up the `IMAGE_ID` (the `kernel-install` in the systemd package is too old, and it only reacts to `IMAGE_ID` in the environment anyway only when `/etc/kernel/entry-token` doesn't exist, which it does for us).

Thinking about this again, it might be cleaner though, to just pass in these variables as the `env` parameter in all the `run_foo_script` functions.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:3210]: Looking at the call sites of `nspaw_params_for_build_sources`, I see that the variables are already set for build scripts, so those lines should be removed then.

Comment (mcassaniti) [reply mkosi/__init__.py:3210]: When I look the call sites are [here](https://github.com/systemd/mkosi/blob/59aa8283d5ce564d5ea7c04103168e4cf35eb552/mkosi/__init__.py#L3179) in `run_prepare_script()` and [here](https://github.com/systemd/mkosi/blob/59aa8283d5ce564d5ea7c04103168e4cf35eb552/mkosi/__init__.py#L7394) in `run_build_script()`. It's the second call site that sets these variables but not the first. I'll remove [these lines](https://github.com/systemd/mkosi/blob/59aa8283d5ce564d5ea7c04103168e4cf35eb552/mkosi/__init__.py#L7388-#L7392) from `run_build_script()` to compensate.

**Comment by mcassaniti** [comment]
I've done a bit of a cleanup as well for `run_build_script()` because there was redundancy between setting the environment variables there and also in `nspawn_params_for_build_sources()`.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:3216]:
```suggestion
def extend_run_env(args: MkosiArgs, env: Mapping[str, str]) -> Dict[str, str]:
```
since it's clear what we return here. I think that will also make the cast further below unnecessary.

Comment (mcassaniti) [reply mkosi/__init__.py]: Fixed. It seems you were right about the additional cast as well.

**Review by behrmann — APPROVED** [review]
LGTM

**Comment by mcassaniti** [comment]
@bluca Rebase done.

**Comment by mcassaniti** [comment]
@bluca, @behrmann, @DaanDeMeyer who wants to merge this one?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Now that MkosiState is a thing, we should do this a little differently.

Similar to MkosiConfig, we should have a field `environment` in MkosiState, that is initialized from `config.environment` + plus any extras we want to add on top, like `IMAGE_ID` and `IMAGE_VERSION`

Then, we should pass `state.environment` instead of `config.environment` everywhere.

**Comment by mcassaniti** [comment]
> Now that MkosiState is a thing, we should do this a little differently.
> 
> Similar to MkosiConfig, we should have a field `environment` in MkosiState, that is initialized from `config.environment` + plus any extras we want to add on top, like `IMAGE_ID` and `IMAGE_VERSION`
> 
> Then, we should pass `state.environment` instead of `config.environment` everywhere.

We found that adding `IMAGE_ID` and `IMAGE_VERSION` to the environment all the time may have ran into issues (somehow) with `debian_gpt4` CI tests under qemu. See [this](https://github.com/systemd/mkosi/pull/1134#issuecomment-1220531516) previous comment requesting only changing specific call sites.

**Comment by DaanDeMeyer** [comment]
That should be fixed after https://github.com/systemd/mkosi/pull/1170

**Comment by mcassaniti** [comment]
I'll wait until #1170 is merged before I update the PR. I'd rather have the CI tests pass. I'm also testing my patches with my own builds and I'd like to confirm they work before I push.

**Comment by mcassaniti** [comment]
Take a look now and see what you think.

File: mkosi/backend.py
Comment (DaanDeMeyer) [inline mkosi/backend.py:618]: No need for this method, we can just pass the environment to the MkosiConfig constructor

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/backend.py
Comment (mcassaniti) [reply mkosi/backend.py]: I believe this is now resolved.

File: mkosi/__init__.py
Comment (DaanDeMeyer) [inline mkosi/__init__.py:6500]: These should not be set in `args.environment`, since then they'll end up in `config.environment`. We only want user config in `config.environment`, not the extra stuff we determine ourselves, so this can't be added to `args.environment`, you need to make sure it only exists in `state.environment` and not in `config.environment`

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: mkosi/__init__.py
Comment (mcassaniti) [reply mkosi/__init__.py:6500]: I'm not sure how you would like this resolved. I could essentially turn that `init_env()` method I wrote before into something that wraps `config.environment` during `MkosiState` instantiation, but I'd like some clarification. By this I mean:

```
state = MkosiState(
    ...
    environment = adjust_env(config),
    ...
)
```

Comment (mcassaniti) [reply mkosi/__init__.py:6500]: I could also use getters and setters, but it's getting a bit ugly.

**Comment by DaanDeMeyer** [comment]
Fixed by #1178

**Pull request closed without merging**
